### PR TITLE
(prometheus) auto complete support

### PR DESCRIPTION
**Pull request by mtanda**
Rebase https://github.com/grafana/grafana/pull/3027 to master.
And add toggle switch.

### CODE REVIEW

File: public/app/plugins/datasource/prometheus/func_editor.js [hunk 1]
```diff
@@ -0,0 +1,166 @@
...
+          function inputBlur() {
+            /* jshint validthis:true */
+            var $input = $(this);
+            ctrl.target.expr = $input.val();
+            ctrl.refreshMetricData();
+          }
+
+          function inputKeyPress(e) {
+            /* jshint validthis:true */
+            if (e.which === 13) {
```
Comment (grobie) [inline public/app/plugins/datasource/prometheus/func_editor.js]: Can we define a variable for this magic number (as there are no constants in js I guess)?

**Comment by grobie** [comment]
I tried to run this, but I couldn't get it to work. Nothing has changed, when I start to type a metric name, I don't get a suggestion list. I could verify that `ng-show="ctrl.target.textEditor"` has been added to the `li` element.

What should I see? What's the best way to debug this?

p.s. There is by the way a leftover console.log in the old metric suggestion. If you click into that metric suggestion field, a `d { ... }` pops up in the console.

**Comment by mtanda** [comment]
@grobie You can toggle auto complete mode by clicking pencil icon.
typeahead

I'll check your feedback. Thanks for testing :-)

**Comment by grobie** [comment]
Hah, I didn't notice that at all. Nice! Seems to work great so far. I'd say let's make this the default and remove the pencil and metric lookup.

It would be good to ensure that the metrics list gets refreshed on server change. We have many different Prometheus servers and it should suggest the right metrics when another one gets selected.

What do you think about status message while loading the metrics list from Prometheus? It could be a short text "Loading auto-suggestions ..." next to the query expression input field.

File: public/app/plugins/datasource/prometheus/func_editor.js (hunk 1 above)
Comment (mtanda) [inline public/app/plugins/datasource/prometheus/func_editor.js]: I don't know well about event code, but Grafana already have such magic number, https://github.com/grafana/grafana/blob/75555c470b695fb196fba7256a6c7b30ecb340a4/public/app/core/directives/metric_segment.js#L131.

And this code is originally written by @juliusv , I just porting the code to Grafana.

**Comment by mtanda** [comment]
> I'd say let's make this the default and remove the pencil and metric lookup.

As I told in comment, this is just porting, so new syntax isn't supported.
This could be very useful as is, but I want to prepare the way to disable this feature.

> It would be good to ensure that the metrics list gets refreshed on server change. We have many 
different Prometheus servers and it should suggest the right metrics when another one gets selected.

To ensure that, need to refresh when datasource is changed in query editor.
I'll investigate how I can do that.

> What do you think about status message while loading the metrics list from Prometheus? It could be a short text "Loading auto-suggestions ..." during load next to the query expression input field.

If loading the metrics take so long time, it is helpful for user.
I'll add it.

**Comment by mtanda** [comment]
> To ensure that, need to refresh when datasource is changed in query editor.

When I switch datasource, relink() called, and refresh the metric list.
I don't need to change the code.

**Comment by mtanda** [comment]
Add loading message and make auto complete mode as default.

**Comment by mtanda** [comment]
Now support on/by/without/group_left/group_right.

**Comment by mtanda** [comment]
@torkelo I notice that you start to refactor Prometheus query editor.
https://github.com/grafana/grafana/commits/query-part-refactor

The Prometheus query editor becomes like Graphite query editor?
Would you consider to support auto complete in refactoring work?

**Comment by torkelo** [comment]
Yes! Want to make the prmetheus editor like the graphite one with autocomplete for metric labels and label values

**Comment by mtanda** [comment]
Wow! It's really fantastic!! 👍 👍 
The UI becomes really cool.
Thanks!

**Comment by mtanda** [comment]
When https://github.com/grafana/grafana/issues/5117 is implemented, I close this.

**Comment by mtanda** [comment]
Now, I think this is quite useful whether https://github.com/grafana/grafana/issues/5117 is implemented or not.
The text editor interface is easier to input if the user is familiar with Prometheus.
Please consider merge this PR.

**Comment by torkelo** [comment]
closing this, will have to be implemented in the new prometheus query editor, https://github.com/grafana/grafana/tree/prometheus_query_editor_v2

**Pull request closed without merging**